src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReferenceHandling.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// This class defines the various ways the <see cref="JsonSerializer"/> can deal with references on serialization and deserialization.
+    /// </summary>
+    public sealed class ReferenceHandling
+    {
+        /// <summary>
+        /// Reference handling semantics will not be applied to complex JSON types when serializing or deserializing.
+        /// </summary>
+        public static ReferenceHandling Default { get; } = new ReferenceHandling(PreserveReferencesHandling.None);
+        /// <summary>
+        /// Reference metadata will be written and honored when deserializing and serializing complex JSON types.
```

Comment:
As part of documentation we should expand on this (maybe as a remark). What does this mean for cycles, and for duplicate references. We should consider adding some parts of the spec here. Same goes with `Default`.

What is metadata? What does `complex JSON types mean`? What happens for structs/value types/etc.? What "equality" is considered?

Reply:
> What does complex JSON types mean?

We can replace that with JSON objects and JSON arrays.

Reply:
How about this:

## Preserve
* On Serialize: When writing complex CLR types (e.g. POCOs/non-primitive types), the serializer also writes metadata (`$id`, `$values`, and `$ref`) properties within them in order to reference them later by writing a pointer to the previously written JSON object or array on the CLR objects that are identical; this is very useful to prevent serialization cycles and preserve reference equality through serialization.

* On Deserialize: The metadata properties emitted on serialization will be expected (although they are not mandatory) and the deserializer will try to understand them.


## Default
* On Serialize: Throw a `JsonException` when `MaxDepth` is exceeded. This may occur by either a reference loop or by passing a very deep object. This option will not affect the performance of the serializer.

* On Deserialize: Metadata properties will not be consumed, therefore metadata properties (`$id`, `$values`, and `$ref`) will be treated as regular properties that can map to a real property using `JsonPropertyName` or be added to the `JsonExtensionData` overflow dictionary.